Support masquerade subnet to be custom configurable as a day2 operation

arghosh93 opened the pull request:
Delete stale masquerade subnet resources if subnet gets changed at day 2
This PR does following:

- Removes following linux resources if masquerade subnet gets changed
  (node side):

  * Removes old V4HostMasqueradeIP and V6HostMasqueradeIP from bridge.

  * Removes stale neighbour entries V4OVNMasqueradeIP, V6OVNMasqueradeIP,
V4DummyNextHopMasqueradeIP and V6DummyNextHopMasqueradeIP if exists.

  * Removes stale masquerade route added by addMasqueradeRoute() function
while starting up the gateway.

  * Removes stale iptables rules created for masquerade subnet based
on ipForwarding and Gateway mode.

- Removes following linux resources if masquerade subnet gets changed
  (ovnkube-controller to NBDB side):

  * Removes logical router static route used by gateway router and
referencing old masquerade subnet.

  * Removes static mac binding for gateway router's rtoe logical port
referencing old masquerade subnet.

Note, the node now sets an annotation to indicate its masquerade subnet
that it last configured. The node uses this at start up to determine if
there has been a change and cleanup is needed.

On the ovnkube-controller side, it also uses this annotation to
determine if the node has changed. However, it may be racy to rely on
this as the node thread may have already updated the annotation by the
time the ovnkube-controller side handles the cleanup. Therefore, in
addition to the annotation ovnkube-controller will additionally scan for
stale routes in NBDB and then derive the route and mac binding to remove
that way. In order to facilitate this, the masquerade route now has an
external_id present (same as the key used in the annotation) to
distinguish which routes are masquerade routes.

Failure to delete things is not usually an overall failure for OVNK.
Therefore upon failing to clean something up, the error is logged, but
startup continues.

Finally, kind.sh and helmet are updated to use a larger masquerade subnet by
default. OVN-Kubernetes defaults themselves remain unchanged.

Co-authored-by: Tim Rozet <trozet@redhat.com>
Signed-off-by: Arnab Ghosh <arnabghosh89@gmail.com>


There is some unit testing included. Also tested on a kind cluster by deploying with default masquerade subnet, then editing daemonset to a new masquerade subnet and checking that annotations and resources were cleaned appropriately.

kyrtapz commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,86 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+	for _, addr := range addrs {
+		if addr.Label == fmt.Sprintf("%s-masq", link.Attrs().Name) && !utilnet.IsIPv6(addr.IP) &&
+			!addr.IPNet.Contains(config.Gateway.MasqueradeIPs.V4OVNMasqueradeIP) {
+			if err = config.AllocateV4MasqueradeIPs(iputils.PrevIP(iputils.PrevIP(addr.IP)), &staleMasqueradeIPs); err != nil {
```
Would it be possible to read the addresses from the link instead of relying on the `AllocateV4MasqueradeIPs` logic?

arghosh93 replied:
From link we only get V4HostMasqueradeIP and we need all other IPs to be derived from that so that we can remove routes, neighbour entries and iptables rules. Can there be any problem if I make allocateV4MasqueradeIPs function public?

kyrtapz commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,86 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+	var staleMasqueradeIPs config.MasqueradeIPsConfig
+	for _, addr := range addrs {
+		if addr.Label == fmt.Sprintf("%s-masq", link.Attrs().Name) && !utilnet.IsIPv6(addr.IP) &&
+			!addr.IPNet.Contains(config.Gateway.MasqueradeIPs.V4OVNMasqueradeIP) {
```
Why `Contains` instead of direct address comparison?

arghosh93 replied:
yaa, makes sense to check equal to V4HostMasqueradeIP instead of Contains.

kyrtapz commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,86 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+	var subnets []*net.IPNet
+	var neighborIPs []net.IP
+	if config.IPv4Mode {
+		_, masqIPNet, _ := net.ParseCIDR(fmt.Sprintf("%s/32", staleMasqueradeIPs.V4OVNMasqueradeIP.String()))
```
Do not ignore the error please.

arghosh93 replied:
Ack

kyrtapz commented on go-controller/pkg/node/gateway_iptables.go:
```diff
@@ -366,6 +366,40 @@ func getGatewayForwardRules(cidrs []*net.IPNet) []nodeipt.Rule {
 }
 
+func getStaleMasqueradeIptablesRules(masqueradeIP net.IP) []nodeipt.Rule {
+	var returnRules []nodeipt.Rule
```
there is no reason to define `returnRules`, why not:
```
return []nodeipt.Rule{
		{
			Table: "filter",
			Chain: "FORWARD",
			Args: []string{
				"-s", masqueradeIP.String(),
				"-j", "ACCEPT",
			},
			Protocol: getIPTablesProtocol(masqueradeIP.String()),
		},
		{
			Table: "filter",
			Chain: "FORWARD",
			Args: []string{
				"-d", masqueradeIP.String(),
				"-j", "ACCEPT",
			},
			Protocol: getIPTablesProtocol(masqueradeIP.String()),
		},
		{
			Table: "nat",
			Chain: "POSTROUTING",
			Args: []string{
				"-s", masqueradeIP.String(),
				"-j", "MASQUERADE",
			},
			Protocol: getIPTablesProtocol(masqueradeIP.String()),
		},
	}
```

arghosh93 replied:
yaa, makes sense

kyrtapz commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,86 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
 	return flows
 }
+
+// deleteStaleMasqueradeResources() function removes stale Linux resources when
```
NIT:
```suggestion
// deleteStaleMasqueradeResources removes stale Linux resources when
```

arghosh93 replied:
Ack

kyrtapz commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,86 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+	return nil
+}
+
+// deleteMasqueradeResources() function removes following Linux resources given a
```
NIT: 
```suggestion
// deleteMasqueradeResources removes following Linux resources given a
```

arghosh93 replied:
Ack

kyrtapz commented on go-controller/pkg/util/net_linux.go:
```diff
@@ -223,6 +223,25 @@ func LinkAddrFlush(link netlink.Link) error {
 	return nil
 }
 
+// LinkAddrDeleteMasqueradeIP() helps to remove stale V4HostMasqueradeIP or
```
```suggestion
// LinkAddrDeleteMasqueradeIP removes stale V4HostMasqueradeIP or...
```
Also, in the context of this function `V4HostMasqueradeIP/V6HostMasqueradeIP` do not mean much. The function removes addresses that have the `<link-name>-masq` label set.

arghosh93 replied:
yes, correct. I will modify the comment.

kyrtapz commented on go-controller/pkg/util/net_linux.go:
```diff
@@ -240,10 +259,21 @@ func LinkAddrExist(link netlink.Link, address *net.IPNet) (bool, error) {
+// This function adds a label '<bridge_name>-masq' while adding either V4HostMasqueradeIP
+// or V6HostMasqueradeIP to a bridge. This is to help identify old masquerade subnet in case it
+// gets changed at day 2.
 func LinkAddrAdd(link netlink.Link, address *net.IPNet, flags, preferredLifetime, validLifetime int) error {
```
Can we just use `netLinkOps.AddrAdd` directly when adding the masq addr instead of hiding it under `LinkAddrAdd`?
Also, it might be useful to extract the label generation into a util function.

arghosh93 commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,86 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+
+	var staleMasqueradeIPs config.MasqueradeIPsConfig
+	for _, addr := range addrs {
+		if addr.Label == fmt.Sprintf("%s-masq", link.Attrs().Name) && !utilnet.IsIPv6(addr.IP) &&
```
Yes, @kyrtapz , I am working on this.

qinqon commented on go-controller/pkg/config/config.go:
```diff
@@ -137,8 +137,8 @@ var (
 		V6JoinSubnet:       "fd98::/64",
-		V4MasqueradeSubnet: "169.254.169.0/29",
-		V6MasqueradeSubnet: "fd69::/125",
+		V4MasqueradeSubnet: "169.254.169.0/16",
```
Agree, first addres for NextIP should be `169.254.0.1`

arghosh93 replied:
yaa, this was a mistake. Thanks for pointing.

qinqon reviewed (changes requested):
- Harcode obsolete masq
- Do proper error wrapping using %w instead of %s or %v.

arghosh93 commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,86 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+// deleteStaleMasqueradeResources() function removes stale Linux resources when
+// config.Gateway.V4MasqueradeSubnet or config.Gateway.V4MasqueradeSubnet gets
+// changed at day 2.
+func deleteStaleMasqueradeResources(bridgeName string) error {
```
@qinqon I was trying hard to not hardcode anything which we are going to obsolete. Instead I preferred cleaning resources after deriving IPs based on the masquerade address attached to br-ex link.
I would request you to check my approach once. If my approach is too dirty then I will go ahead and hardcode the old masquerade subnet.
I will push local changes soon.

qinqon replied:
An alternative can be using different addr label for the new masquerade like `breth0-masq-1` so you can search for the old label `breth0-masq` instead of calculating stuff.

Also if users change configuration you are not going to be able to calculate the old one.

arghosh93 replied:
Resolving this conversation as I have moved away from adding label to address on external bridge due to iproute not supporting label on IPv6 addresses.

arghosh93 commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -1769,6 +1770,14 @@ func newSharedGateway(nodeName string, subnets []*net.IPNet, gwNextHops []net.IP
+		// gets changed as part of day2 operation. Need to run before setNodeMasqueradeIPOnExtBridge()
+		// while setting up the gateway.
+		if err := deleteStaleMasqueradeResources(gwBridge.bridgeName); err != nil {
+			return fmt.Errorf("failed to remove stale masquerade resources: %s", err)
```
Ack

arghosh93 commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2069,6 +2078,10 @@ func setNodeMasqueradeIPOnExtBridge(extBridgeName string) error {
 		if exists, err := util.LinkAddrExist(extBridge, bridgeCIDR.ipNet); err == nil && !exists {
+			// Delete any stale masquerade IP from bridge before adding new one.
+			if err := util.LinkAddrDeleteMasqueradeIP(extBridge, bridgeCIDR.ipNet); err != nil {
+				return fmt.Errorf("failed to remove masquerade IP from bridge %s: %s", extBridge, err)
```
Ack

arghosh93 commented:
Added couple of extra functions to take care of following things:

 * Remove logical router static route used by gateway router referencing old masquerade subnet.
 * Remove static mac binding for gateway router's rtoe logical port referencing old masquerade subnet.
 * Modified couple of test cases using IP from current default masquerade subnet.

qinqon commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2173,3 +2186,110 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+		// dummy label on already retrived address if its IP exists in current default masquerade subnet.
+		// This allows matching further if conditions in this function and finally delete stale linux
+		// resources(route, neighbour, iptables rules) for old default masquerade subnet.
+		if v4MasqueradeCIDR.Contains(addr.IPNet.IP) || v6MasqueradeCIDR.Contains(addr.IPNet.IP) {
```
But didn't we have the label already on those addresses ?

arghosh93 commented:
[FAIL] Services [It] Creates a host-network service, and ensures that host-network pods can connect to it
--> Passing locally.

[FAIL] [sig-network] Networking Granular Checks: Services [It] should function for service endpoints using hostNetwork [sig-network]
--> Seems 1 node never became ready and curl to a host-networked POD on not ready node failed.
~~~
[FAILED] failed dialing endpoint, did not find expected responses... 
  Tries 39
  Command curl -g -q -s 'http://10.244.1.153:9080/dial?request=hostname&protocol=http&host=172.18.0.3&port=30920&tries=1'
  Jul 12 16:33:46.611: INFO: Waiting up to 7m0s for all (but 0) nodes to be ready
  retrieved map[ovn-worker:{} ovn-worker2:{}]
  expected map[ovn-control-plane:{} ovn-worker:{} ovn-worker2:{}]
~~~

[Upgrade OVN from Master to PR branch based image (local)]
--> It seems it timed out. It did not fail in last run(https://github.com/ovn-org/ovn-kubernetes/actions/runs/9892290667).  After that I only made changes to masquerade subnet in an e2e test. 

 [FAIL] External Gateway With Admin Policy Based External Route CRs e2e non-vxlan external gateway through a gateway pod Should validate ICMP connectivity to an external gateway's loopback address via a gateway pod [It] ipv4
--> It did not fail in last run(https://github.com/ovn-org/ovn-kubernetes/actions/runs/9892290667/job/27326085598).

arghosh93 commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -1373,3 +1388,172 @@ func configureGlobalForwarding() error {
+			if !v4ClusterSubnet.Contains(route.Dst.IP) && !utilnet.IsIPv6CIDR(route.Dst) &&
+				(!route.Dst.IP.Equal(config.Gateway.MasqueradeIPs.V4HostETPLocalMasqueradeIP) ||
+					(route.Dst.IP.Equal(config.Gateway.MasqueradeIPs.V4HostETPLocalMasqueradeIP) && !v4RouteExists)) {
+				if err := config.AllocateV4MasqueradeIPs(ip.PrevIP(ip.PrevIP(ip.PrevIP(route.Dst.IP))), staleMasqueradeIPs); err != nil {
```
if v4RouteExists, then most certainly there wont be any stale masquerade resources. I am checking routes on external bridge only to make sure to delete stale masquerade resources when only prefix length changes.
If I do it like below then I would need to go through mp0 routes differently for v4 and v6.

`if !v4RouteExists {
		for _, route := range mp0Routes {
			for _, v4ClusterSubnet := range v4ClusterSubnets {
				if !v4ClusterSubnet.Contains(route.Dst.IP) && !utilnet.IsIPv6CIDR(route.Dst) &&
				!route.Dst.IP.Equal(config.Gateway.MasqueradeIPs.V4HostETPLocalMasqueradeIP) {
					if err := config.AllocateV4MasqueradeIPs(ip.PrevIP(ip.PrevIP(ip.PrevIP(route.Dst.IP))), staleMasqueradeIPs); err != nil {
						return fmt.Errorf("unable to determine stale V4MasqueradeIPs: %s", err)
					}
				}
			}
		}
	}`

arghosh93 commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -1373,3 +1388,172 @@ func configureGlobalForwarding() error {
+		for _, v4ClusterSubnet := range v4ClusterSubnets {
+			if !v4ClusterSubnet.Contains(route.Dst.IP) && !utilnet.IsIPv6CIDR(route.Dst) &&
+				(!route.Dst.IP.Equal(config.Gateway.MasqueradeIPs.V4HostETPLocalMasqueradeIP) ||
+					(route.Dst.IP.Equal(config.Gateway.MasqueradeIPs.V4HostETPLocalMasqueradeIP) && !v4RouteExists)) {
```
So, in this logic I am relying on 2 routes:
169.254.169.3 via 10.128.2.1 dev ovn-k8s-mp0  --> this route alone is not sufficient to identify stale masquerade subnet as HostETPLocalMasqueradeIP will remain same if only prefix length changes but IPNet.IP remains same.

169.254.169.0/29 dev br-ex proto kernel scope link src 169.254.169.2  --> This route on external bridge will tell us masquerade IPNet.

I am checking following to decide if there are any stale resources:
1. HostETPLocalMasqueradeIP on mp0 route is not equal to current HostETPLocalMasqueradeIP IP.
2. If HostETPLocalMasqueradeIP is equal to current HostETPLocalMasqueradeIP then masquerade subnet(IPNet) on external bridge also must not be equal to current config.Gateway.V4MasqueradeSubnet. 

When above conditions are met, I am considering the IP as a stale masquerade IP then trying to calculate all other masquerade IPs by using AllocateV4MasqueradeIPs function.

arghosh93 commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -1373,3 +1388,172 @@ func configureGlobalForwarding() error {
+	//   This is to make sure we remove stale masquerade resources when only prefix length changes.
+	v4ClusterSubnets, v6ClusterSubnets := util.GetClusterSubnets()
+
+	for _, route := range routes {
```
I may have a conceptual gap here but cant IPv6 link local addresses be used as masquerade subnet just like IPv4?

arghosh93 commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -1373,3 +1388,172 @@ func configureGlobalForwarding() error {
+		LinkIndex: link.Attrs().Index,
+	}
+	filterMask := netlink.RT_FILTER_OIF
+	routes, err := util.GetNetLinkOps().RouteListFiltered(netlink.FAMILY_ALL, routeFilter, filterMask)
```
Ack

arghosh93 commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -1373,3 +1388,172 @@ func configureGlobalForwarding() error {
+
+// getOldMasquerdeSubnet determines the old masquerade subnet if it has changed at day 2.
+// It evaluates routes for ovn-k8s-mp0 interface to determine old subnet.
+func getOldMasquerdeSubnet(staleMasqueradeIPs *config.MasqueradeIPsConfig, bridgeName string) error {
```
I thought about this approach before and would have been easier to implement but moved away due to following reasons:

- We are changing default masquerade subnet from 169.254.169.0/29 to 169.254.0.0/16 and during first upgrade from a version using old default masquerade subnet to a version using new default masquerade subnet we would need to remove stale entries and during that time there wont be any annotations with old default masquerade subnet. 

- Users may have customized masquerade subnet during cluster installation. From downstream prospective, we do support custom masquerade subnet either during installation or during SDN to OVN migration. So, we cant also assume that 169.254.169.0/29 masquerade subnet is being used in all old clusters.

arghosh93 commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -820,6 +821,20 @@ func (nc *DefaultNodeNetworkController) Start(ctx context.Context) error {
+		return fmt.Errorf("failed to get gateway bridge name: %w", err)
+	}
+
+	if err := deleteStaleMasqueradeResources(bridgeName); err != nil {
```
Yes, initGateway runs later and thats the reason I could not use nc.gateway.GetGatewayBridgeIface() to get bridge name. Instead I am using methods used in initGateway or initGatewayDPUHost to get the bridge name.

The gateway bridge gets created while ovs-configuration service starts up and then OVN PODs run and hence we are not getting link not found error.
For mp0 interface I am checking util.GetNetLinkOps().IsLinkNotFoundError(err) after running util.GetNetLinkOps().LinkByName(types.K8sMgmtIntfName) and returning from the function if link not found.

trozet replied:
But in KIND/upstream there is no ovs-configuration. The gateway code will find the default route, create the bridge, and move the interface into the bridge. All of that happens after your code I think.

arghosh93 commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -1373,3 +1388,172 @@ func configureGlobalForwarding() error {
+// - masquerade route added by addMasqueradeRoute function while starting up the gateway.
+// - iptables rules created for masquerade subnet based on ipForwarding and Gateway mode.
+// - stale HostMasqueradeIP address from gateway bridge
+func deleteMasqueradeResources(link netlink.Link, staleMasqueradeIPs *config.MasqueradeIPsConfig) error {
```
createNodeManagementPorts function flushes routes and link address for ovn-k8s-mp0 interface while starting defaultNodeNetworkController each time. That is the reason I had to run deleteStaleMasqueradeResources before createNodeManagementPorts function gets executed.

arghosh93 commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -781,3 +790,113 @@ func (oc *DefaultNetworkController) deletePolicyBasedRoutes(policyID, priority s
+	if err != nil {
+		return fmt.Errorf("failed to get stale masquerade static route UUID: %s", err)
+	}
+	if len(uuid) > 1 {
```
Ack

arghosh93 commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -781,3 +790,113 @@ func (oc *DefaultNetworkController) deletePolicyBasedRoutes(policyID, priority s
+			return nil, fmt.Errorf("failed to parse static route IPPrefix while removing stale masquerade data: %s", err)
+		}
+
+		// Extract the static route UUID when IPPrefix does not matches all IPs(default gateway)
```
Thanks :)

arghosh93 commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -781,3 +790,113 @@ func (oc *DefaultNetworkController) deletePolicyBasedRoutes(policyID, priority s
+		if err != nil {
+			return fmt.Errorf("failed to list routes by uuid while cleaning up static MAC binding: %s", err)
+		}
+		// Delete satic MAC binding only if Nexthop of static route is not same as dummy masquerade IP
```
Ack

kyrtapz commented on contrib/kind.sh:
```diff
@@ -543,8 +543,8 @@ set_default_params() {
   SVC_CIDR_IPV6=${SVC_CIDR_IPV6:-fd00:10:96::/112}
   JOIN_SUBNET_IPV4=${JOIN_SUBNET_IPV4:-100.64.0.0/16}
   JOIN_SUBNET_IPV6=${JOIN_SUBNET_IPV6:-fd98::/64}
-  MASQUERADE_SUBNET_IPV4=${MASQUERADE_SUBNET_IPV4:-169.254.169.0/29}
```
NIT: In the future, please extract renames into a separate commit. It makes reviewing the other changes easier.

arghosh93 replied:
I will keep this suggestion in mind , Thanks :)

kyrtapz commented on go-controller/pkg/node/default_node_network_controller.go:
```diff
@@ -1373,3 +1388,172 @@ func configureGlobalForwarding() error {
+}
+
+// getOldMasquerdeSubnet determines the old masquerade subnet if it has changed at day 2.
+// It evaluates routes for ovn-k8s-mp0 interface to determine old subnet.
```
So the old masq subnet is getting extracted from a route? Didn't we discuss looking it up in OVN? https://redhat-internal.slack.com/archives/C077NGF5HE1/p1720531355364399?thread_ts=1720529597.653169&cid=C077NGF5HE1

arghosh93 replied:
I evaluated that option but DefaultNodeNetworkController do not have nbClient.

arghosh93 replied:
resolving all conversations due to change in approach.

kyrtapz commented on go-controller/pkg/node/gateway_init.go:
```diff
@@ -563,3 +563,26 @@ func (nc *DefaultNodeNetworkController) updateGatewayMAC(link netlink.Link) erro
+	if err != nil {
+		return "", err
+	}
+	ifAddrs, err := getNetworkInterfaceIPAddresses(gatewayIntf)
```
is this required? I don't think it is getting used as we only return the bridge name

arghosh93 replied:
yes, you are right. We dont need this it seems. I will test and update the PR accordingly. Thanks for such a nice suggestion :)

arghosh93 replied:
closing due to change in approach.

kyrtapz commented on go-controller/pkg/node/gateway_init.go:
```diff
@@ -563,3 +563,26 @@ func (nc *DefaultNodeNetworkController) updateGatewayMAC(link netlink.Link) erro
+		if err != nil {
+			return "", err
+		}
+		config.Gateway.Interface = gwIntf
```
Why does a get function modify the global config?

kyrtapz replied:
Thanks for the link, I still don't understand why is that needed in `getGatewayBridgeName`

arghosh93 replied:
getGatewayNextHops function uses config.Gateway.Interface. I think we can expect gateway interface name to be different in case of dpu-host. 
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_init.go#L155-L161 

I agree that I do not have a concrete answer to your question. @trozet might know.

arghosh93 commented on go-controller/pkg/node/gateway_iptables.go:
```diff
@@ -366,11 +366,44 @@ func getGatewayForwardRules(cidrs []*net.IPNet) []nodeipt.Rule {
 
+// getStaleMasqueradeIptablesRules returns all iptables rules may get added for a given masquerade IP.
+func getStaleMasqueradeIptablesRules(masqueradeIP net.IP) []nodeipt.Rule {
+	return []nodeipt.Rule{
```
Ack

trozet commented:
@arghosh93 so @kyrtapz and I spoke for a while today about this PR. We both don't like the mechanism of trying to guess what the old masquerade IP is. It also isn't easy to introduce something into OVNK to hold the state, but also give the ability to change the masquerade subnet in the same PR. We ended up agreeing on the following proposal. Let me know what you think.

1. OVNK introduces new config arguments "v4StaleMasqueradeSubnet", "v6StaleMasqueradeSubnet"
2. When OVNK is started it uses these values to clean any of the linux host config or NBDB config if necessary.
3. An upstream user would need to provide these values when they are migrating the MasqueradeSubnet to something else.
4. In CNO/OCP, CNO maintains the last applied config. When a user goes to change the masquerade subnet, CNO would automatically pass the previous values as stale values for clean up to OVNK.
5. We should document in OCP that a user should not change the value of the masquerade subnet again, until all of the OVNK pods have rolled out. This avoids a state of incomplete cleanup where the n-2 stale masquerade is now lost and OVNK cannot clean it up. Additionally, we should document that worst case scenario a user can just reboot their nodes to clean up stale masquerade config.

With this proposal we don't have to do any guessing in OVNK, and it seems much more explicit.

arghosh93 commented:
I also did not likke the guess work much but was running out of ideas. Initially I tried to attach label to masquerade address in gateway bridge and it worked perfectly fine for IPv4 addresses but later got to know that label on IPv6 addresses are not supported by iproute and had to change approach.

Thanks @trozet @kyrtapz for working out on a plan. This plan sounds good to me. Allow me some time to work on it and I will send new changes as soon as possible. I will also raise questions if I have any. Thanks once again!

trozet commented:
@arghosh93 thanks. I also just realized we should default "v4StaleMasqueradeSubnet", "v6StaleMasqueradeSubnet"  to the old masquerade values (169.254.169.0/29, fd69::/125). That way when someone upgrades who has never overridden the defaults, we will clean up the old stuff as you specify new defaults for V4MasqueradeSubnet, and V6MasqueradeSubnet

arghosh93 commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,9 @@ func (oc *DefaultNetworkController) gatewayInit(nodeName string, clusterIPSubnet
 			Nexthop:    nextHop.String(),
 			OutputPort: &externalRouterPort,
+			ExternalIDs: map[string]string{
+				"subnet": "masquerade",
```
closing this conversation due to change in approach.

arghosh93 commented:
@trozet @kyrtapz Updated both https://github.com/openshift/cluster-network-operator/pull/2421 and this PR as per discussion. Thanks again for your valuable suggestions.

maiqueb commented:
@arghosh93 o/

Is this PR in a good state ? I am considering rebasing it, so I could work on top of it.

arghosh93 commented:
yes @maiqueb , I am waiting on reviews from Tim and Patryk though. However I dont expect major changes(fingers crossed :D).

trozet reviewed (commented):
overall looks good, just one thing to get clarity on

kyrtapz commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2158,3 +2166,106 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+				return fmt.Errorf("failed to remove masquerade IP from bridge %s: %w", link, err)
+			}
+		} else {
+			// This is to make sure that we don't delete anything if stale HostMasqueradeIP does not
```
Instead of doing an early exit on errors/missing entries can we go through all cleanup steps and return one aggregated error? This would ensure that we would at least try to cleanup everything.

kyrtapz replied:
How does returning nil ensures we are not removing user configured resources? This function assumes we own all of the resources that are being removed.

arghosh93 replied:
I think I should get rid of this else condition. another downfall of this is that IPv6 masquerade resource cleanup may get hampered due to this return when IPv4 masquerade link address does not exist. 

So, I am going to aggregate all errors instead of returning immediately and going to log a warning message as Tim suggested.

arghosh93 replied:
on a second thought it seems it would be better to have two different functions to delete v4 masquerade stale resources and v6 masquerade stale resources. I think its better to not delete some resources instead of deleting resources which user might have configured.
deleteV4MasqueradeResources & deleteV6MasqueradeResources: 
- If masquerade IP does not exist on link then dont remove other linux resources and just return nil error.
- Log warning message to notify user that some linux resources from stale masquerade subnet might still be there.
- Use aggregate type error so that the function does not return immediately upon a single failure while deleting resource.

Does this sound an acceptable option?

trozet replied:
masquerade IP does not exist on the *mp0* link right? I think this is preferable to me.

I'm a bit worried about this case:
1. A user is already using OVNK before this patch, and they use 169.254.169.x as a secondary IP address on their br-ex(breth0) interface already, so they were overriding masquerade subnet with something else.
2. The user upgrades to this new version, and we wipe his config due to the default value in v4StaleMasquerade.

I think by checking to see is the ip first on mp0, that gives us a signal that ok this was configured as a masquerade subnet before by ovnk, and gives us some validation that it should be ok for us to remove things related to that subnet. I think if it isn't there we cannot be sure that we should remove config and should log a warning. Its ok to do it independently for v4 and v6. I'm not sure I like the idea of propagating an error, unless it is just logged.

kyrtapz commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2158,3 +2166,106 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+
+	for _, ip := range neighborIPs {
+		if err := util.LinkNeighDel(link, ip); err != nil {
+			klog.Warningf("Failed to remove IP neighbour entry for ip %s, on iface %s: %v",
```
Why are we just loging a warning here and return an error for all other del operations?

arghosh93 replied:
agh, sorry sorry. I should return error here too. Thanks so much Patryk.

kyrtapz commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -781,3 +787,58 @@ func (oc *DefaultNetworkController) deletePolicyBasedRoutes(policyID, priority s
+			return item.OutputPort != nil && *item.OutputPort == logicalport &&
+				item.IPPrefix == prefix && utilnet.IPFamilyOfCIDRString(item.IPPrefix) == utilnet.IPFamilyOf(nextHop)
+		}
+		if err := libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicate(nbClient, routerName, staticRoutePredicate); err != nil {
```
This function could be refactored to do just one nbdb transaction.
`for _, nextHop := range nextHopIPs {` could be included in the predicates and the Delete operations could be joined.

arghosh93 replied:
Yes, I agree, delete operations can be joined but it looks cleaner to me now. should I still go ahead and combine both delete operationModels ?

kyrtapz replied:
Combining all operations into one transaction means there is less chance for a partial cleanup and results in less calls to the db.
 I am ok to keep it as is if that's what you prefer, it shouldn't break anything :)

kyrtapz commented:
It doesn't have to be done as part of this PR but it would be great if we had an e2e test for changing the masquerade subnet day2.

arghosh93 commented on go-controller/pkg/node/gateway_init.go:
```diff
@@ -457,6 +457,13 @@ func (nc *DefaultNodeNetworkController) initGatewayDPUHost(kubeNodeIP net.IP) er
 	}
 
+	// Delete stale masquerade resources if there are any. This is to make sure that there
+	// are no Linux resouces with IP from old masquerade subnet when masquerade subnet
```
Thanks !

trozet commented:
looks like some flakes are happening:
Granular Checks: Services [38;5;9m[1m[It] should function for service endpoints using hostNetwork
https://github.com/ovn-org/ovn-kubernetes/actions/runs/10224345092/job/28292319498?pr=4496

looking into it

trozet commented:
The jobs flaking are ones with ic-disabled. That's because on the ovnkube-node pod it is still getting the old subnet of 169.254.169.0/29, while master is getting the new subnet of 169.254.0.0/16. This is really odd, because the ovnkube-node pod yaml has the environment variable set:

```
trozet@fedora:~/go/src/github.com/ovn-org/ovn-kubernetes/dist/yaml$ oc exec -n ovn-kubernetes -it  ovnkube-node-cpqfq -c ovnkube-node /bin/bash
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
[root@ovn-worker2 ~]# echo $OVN_V4_MASQUERADE_SUBNET
169.254.0.0/16
```

but in the ovnkube-node start up, i can see the cli config never gets the value:
```
ovn_conntrack_zone_flag=--conntrack-zone=64000
=============== ovn-node   --init-node
info: Waiting for process_ready ovnkube to come up, waiting 1s ...
I0803 17:25:09.579887     930 config.go:1811] TROZET: config from file: 169.254.169.0/29
I0803 17:25:09.580046     930 config.go:1815] TROZET: config after file: 169.254.169.0/29
I0803 17:25:09.580052     930 config.go:1826] TROZET: config from cli: 169.254.169.0/29
I0803 17:25:09.580065     930 config.go:1831] TROZET: config after cli: 169.254.169.0/29

I0803 17:25:09.580242     930 config.go:2320] Gateway config: {Mode:shared Interface: EgressGWInterface: NextHop: VLANID:0 NodeportEnable:true DisableSNATMultipleGWs:false V4JoinSubnet:100.64.0.0/16 V6JoinSubnet:fd98::/64 V4MasqueradeSubnet:169.254.169.0/29 V6MasqueradeSubnet:fd69::/125 MasqueradeIPs:{V4OVNMasqueradeIP:169.254.169.1 V6OVNMasqueradeIP:fd69::1 V4HostMasqueradeIP:169.254.169.2 V6HostMasqueradeIP:fd69::2 V4HostETPLocalMasqueradeIP:169.254.169.3 V6HostETPLocalMasqueradeIP:fd69::3 V4DummyNextHopMasqueradeIP:169.254.169.4 V6DummyNextHopMasqueradeIP:fd69::4 V4OVNServiceHairpinMasqueradeIP:169.254.169.5 V6OVNServiceHairpinMasqueradeIP:fd69::5} DisablePacketMTUCheck:false RouterSubnet: SingleNode:false DisableForwarding:false AllowNoUplink:false}
```

really confused how this is possible, unless the cli parser is bailing and not reporting error when it fails to parse the environment variables

trozet commented:
Fixed the problem, it was missing being set in ovnkube.sh for ovnkube-node.

trozet commented:
Looks like some flake in hybrid overlay tests:

```
2024-08-04T16:38:12.0939360Z [91m[1m• Failure [2.602 seconds][0m
2024-08-04T16:38:12.0939966Z Hybrid SDN Master Operations
2024-08-04T16:38:12.0941135Z [90m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/hybrid_test.go:132[0m
2024-08-04T16:38:12.0942540Z   [91m[1mhandles a HO node is switched to a OVN node [It][0m
2024-08-04T16:38:12.0944090Z   [90m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/hybrid_test.go:1216[0m
2024-08-04T16:38:12.0944990Z 
2024-08-04T16:38:12.0945270Z   [91mTimed out after 2.001s.
2024-08-04T16:38:12.0945768Z   Expected
2024-08-04T16:38:12.0946367Z       <[]*nbdb.LogicalRouterPolicy | len:1, cap:1>: [
2024-08-04T16:38:12.0947031Z           {
2024-08-04T16:38:12.0947870Z               UUID: "ac4ad1bc-2680-4c98-a279-4938e16e0496",
2024-08-04T16:38:12.0948651Z               Action: "reroute",
2024-08-04T16:38:12.0949292Z               BFDSessions: nil,
2024-08-04T16:38:12.0949901Z               ExternalIDs: {
2024-08-04T16:38:12.0950797Z                   "name": "hybrid-subnet-node1-gr",
2024-08-04T16:38:12.0951668Z               },
2024-08-04T16:38:12.0952514Z               Match: "ip4.src == 100.64.0.2 && ip4.dst == 10.1.3.0/24",
2024-08-04T16:38:12.0953316Z               Nexthop: nil,
2024-08-04T16:38:12.0953946Z               Nexthops: ["10.1.1.3"],
2024-08-04T16:38:12.0954583Z               Options: nil,
2024-08-04T16:38:12.0955307Z               Priority: 1002,
2024-08-04T16:38:12.0955807Z           },
2024-08-04T16:38:12.0956171Z       ]
2024-08-04T16:38:12.0956606Z   to have length 0[0m
2024-08-04T16:38:12.0956892Z 
2024-08-04T16:38:12.0957722Z   /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/hybrid_test.go:1392
```

https://github.com/ovn-org/ovn-kubernetes/actions/runs/10237683879/job/28321286598?pr=4496

kyrtapz commented on go-controller/pkg/node/gateway_init.go:
```diff
@@ -457,6 +457,13 @@ func (nc *DefaultNodeNetworkController) initGatewayDPUHost(kubeNodeIP net.IP) er
+	// are no Linux resources with IP from old masquerade subnet when masquerade subnet
+	// gets changed as part of day2 operation.
+	if err := deleteStaleMasqueradeResources(gwIntf, nc.name, nc.watchFactory); err != nil {
+		return fmt.Errorf("failed to remove stale masquerade resources: %w", err)
```
Just to confirm, we want to stick to a hard fail on cleanup issues?

trozet replied:
so the way it is written is so that an unexpected error, like failure to parse the kapi node, or unable to calculate the masquerade ips from the annotation value will result in an error. A failure to clean up the resources will only log inside of the deleteStaleMasqueradeResources

kyrtapz replied:
Ah: https://github.com/ovn-org/ovn-kubernetes/pull/4496/files#diff-d3aa58d9b58a0a09264f072df46ab01d0501eb508c4656411ae2dc1ac68fb3c4R2233
thanks! that makes sense

kyrtapz reviewed (approved):
Look good to me, few minor comments

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -2158,3 +2183,136 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
+			return fmt.Errorf("unable to determine stale V6MasqueradeIPs: %s", err)
+		}
+	}
+	link, err := util.LinkSetUp(bridgeName)
```
will fix

kyrtapz reviewed (approved):
/lgtm
